File: 백준/Gold/2110. 공유기 설치/공유기 설치.py
import sys
input = sys.stdin.readline

n, c = map(int, input().split())
x = []
res = 0

for _ in range(n):
    x.append(int(input()))

x.sort()

s = 1
e = x[-1] - x[0]


while s <= e:
    mid = (s + e) // 2
    cur = x[0]
    cnt = 1
    
    for i in range(1, len(x)):
        if x[i] >= cur + mid:
            cnt += 1
            cur = x[i]
    if cnt >= c:
        res = mid
        s = mid+1
    else:
        e = mid-1
print(res)

File: 백준/Gold/2110. 공유기 설치/공유기 설치.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        int n = Integer.parseInt(st.nextToken());
        int c = Integer.parseInt(st.nextToken());
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }
        Arrays.sort(arr);
        long s = 1;
        long e = arr[n-1];
        long res = 0;
        while (s <= e) {
            long total = 0;
            int wifiCnt = 1;
            long mid = (s + e) / 2;
            for (int i = 1; i < n; i++) {
                long dist = arr[i] - arr[i-1];
                total += dist;
                if (total >= mid) {
                    wifiCnt++;
                    total = 0;
                }
            }

            if (wifiCnt >= c) {
                res = Math.max(res, mid);
                s = mid + 1;
            } else {
                e = mid - 1;
            }
        }

        System.out.println(res);
    }
}